Implement WebAssembly optimizer

pdubroy opened the pull request:
A step towards the design described in https://github.com/avibryant/kombu/wiki/Computation-Lifecycle:

- Renamed Wasm `evaluator` function to `optimizer`
- reimplemented the simple `optimize` loop in Wasm
- cache works as described in the Computation Lifecycle doc: unconditionally write to in on the first use of a value, unconditionally read on subsequent uses

avibryant commented on src/core/construct.ts:
```diff
@@ -24,7 +24,7 @@ export function param(name: string, value: number = 1): t.Param {
 export function constant(n: number): t.Constant {
   const bounds = o.constantBounds(n)
-  return { type: t.NumType.Constant, value: n, bounds }
+  return { type: t.NumType.Constant, id: nextID(), value: n, bounds }
```
not a big deal, but do we need constants to have an id? it surprises me if we do.

pdubroy replied:
Fixed in the latest commit.

avibryant reviewed (commented):
I haven't read through the codegen in detail but it's well compartmentalized and I can get you to walk me through it at some later point.

In terms of the outside API this looks good, modulo the one nit about whether constants need ids.

pdubroy commented:
@avibryant I've addressed all TODOs that I originally listed in the PR description, and also removed the `id` field from `Constant`. From my side this is ready to merge.

pdubroy commented:
Gonna use an optimistic merge strategy 😆. Happy to address any comments you have in a follow-up.
